### PR TITLE
noswizzle

### PR DESCRIPTION
**Pull request by jkub**
This does three things:
- narrows the interface to SwizzleablePointer
- Adds one level of type indirection (ArtPointer)
- Adds some comments to SwizzleablePointer implementation

Combined, these make it easier to experiment with and maintain alternate implementations of this pointer type.

### CODE REVIEW

**Comment by Tishj** [comment]
LGTM
I like the separation between SwizzleablePointer and ARTPointer, the extra asserts and comments are valid.
I'm not entirely sure about `empty` in this context, as it doesn't make sense for a pointer to be empty, maybe `valid` would be a better word, then `!pointer.empty()` can turn into `pointer.valid()`

Also slight nitpick: maybe `ARTPointer` over `ArtPointer` ?

File: src/execution/index/art/swizzleable_pointer.cpp [hunk 1]
```diff
@@ -18,15 +18,19 @@ SwizzleablePointer::SwizzleablePointer(duckdb::MetaBlockReader &reader) {
...
 	pointer += offset;
 	// Set the left most bit to indicate this is a swizzled pointer and send it back to the mother-ship
 	uint64_t mask = 1;
 	mask = mask << (pointer_size - 1);
+	// This assumes the 33rd most significant bit of the block_id is zero.
 	pointer |= mask;
 }
 
 SwizzleablePointer &SwizzleablePointer::operator=(const Node *ptr) {
+	// Note: If the object already has a non-swizzled pointer, this will leak memory.
```
Comment (pdet) [inline src/execution/index/art/swizzleable_pointer.cpp:33]: Maybe add an assertion here?

File: src/execution/index/art/swizzleable_pointer.cpp [hunk 2]
```diff
@@ -42,6 +46,8 @@ bool operator!=(const SwizzleablePointer &s_ptr, const uint64_t &ptr) {
 BlockPointer SwizzleablePointer::GetSwizzledBlockInfo() {
 	D_ASSERT(IsSwizzled());
 	idx_t pointer_size = sizeof(pointer) * 8;
+	// Note: This is destructive. Pointer will be invalid after this
+	// operation.
```
Comment (pdet) [inline src/execution/index/art/swizzleable_pointer.cpp:50]: I don't think we use `Note: ` in our comments, just have the comment itself.

**Review by pdet — CHANGES_REQUESTED** [review]
Hi @jkub thanks for your contribution!

I have some small requests.

In additions to the ones I inlined in the code, I think generally speaking comments like
"This assumes" should be replaced with assertions that check these assumptions. This is generally more helpful because if something breaks it becomes immediately obvious where/why.

**Comment by jkub** [comment]
@Tishj : What about implementing "operator bool" rather than valid(), to follow standard C++ pointer idioms?

**Comment by jkub** [comment]
I don't know why "Tidy Check (pull_request)" is failing to find the ARTPointer definition. The compiler doesn't seem to have an issue.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**